README.md
[![Build Status](https://travis-ci.org/seankross/minimap.svg?branch=master)](https://travis-ci.org/seankross/minimap)
[![codecov.io](https://codecov.io/github/seankross/minimap/coverage.svg?branch=master)](https://codecov.io/github/seankross/minimap?branch=master)
[![CRAN version](http://www.r-pkg.org/badges/version/minimap)](https://cran.r-project.org/web/packages/minimap/index.html)

# minimap

A beautiful tile grid map is only a function call away! Tile grid maps are a
great way to display geographic data when you want to represent regions with
equal visual space. No shapefiles required, just supply a vector of postal 
abbreviations and a corresponding vector of colors. Built with base graphics.

If you're unfamilar with tile grid maps, check out the examples below or see the
following news articles:

- [The New York Times](http://www.nytimes.com/interactive/2015/03/04/us/gay-marriage-state-by-state.html)
- [The Washington Post](https://www.washingtonpost.com/graphics/national/minimum-wage/)
- [National Public Radio](http://blog.apps.npr.org/2015/05/11/hex-tile-maps.html)

## Install

```r
install.packages("minimap")
```

## Demos

```r
# La Patria Es Primero

library(minimap)
minimexico(mexico_abb, colorRampPalette(c("#006847", "white", "#CE1126"))(32), 
            estados_name_colors = rep("black", 32))
```

![Mexico](https://raw.githubusercontent.com/seankross/minimap/gh-pages/images/mexico.png)

```r
# Legal Status of Same Sex Marriage in the United States (2008)

library(minimap)

determine_color <- function(status){
  if(status == "bbs")
    "#FFE597"
  else if(status == "nl")
    "#F1F1F0"
  else if (status == "dis")
    "#D0C7B9"
  else if(status == "bbca")
    "#FDC471"
  else
    "#817972"
}

ssm$color <- as.character(sapply(ssm$Status, determine_color))
ssm_2008 <- ssm[ssm$Year == 2008,]
miniusa(ssm_2008$State, state_colors = ssm_2008$color, state_names = TRUE,
        state_name_colors = rep("white", 51))
title(main = "Legal Status of Same Sex Marriage in 2008", line = -1)
```

![USA](https://raw.githubusercontent.com/seankross/minimap/gh-pages/images/usa.png)

```r
# Legal Status of Gay Marriage Over Time

library(minimap)

old_mai <- par()$mai
par(mai = c(0, 0, .75, .5), mfrow = c(2, 4))

for(i in 2008:2015){
  one_year <- ssm[ssm$Year == i,]
  miniusa(one_year$State, state_colors = one_year$color, state_names = FALSE)
  title(main = i, line = -2)
}

mtext("Legal Status of Gay Marriage Over Time", outer = TRUE, side = 3, line = -2)

par(mai = old_mai, mfrow = c(1, 1))
```

![USA](https://raw.githubusercontent.com/seankross/minimap/gh-pages/images/usam.png)

```r
# Forty Years of Canadian Milk Production

library(minimap)
library(dplyr)
library(RColorBrewer)
library(animation)

bin <- function(x){
  qs <- as.numeric(quantile(x, seq(0, 1, .1)))
  sapply(x, function(y){
    which(abs(qs-y)==min(abs(qs-y)))
  })
}

milk_year <- milk %>%
  group_by(Region, Year) %>%
  summarise(Total_kL = sum(Kiloliters)) 
  
max_milk <- milk_year %>%
  group_by(Region) %>%
  summarise(Max = max(Total_kL))
  
milk_year <- milk_year %>%
  left_join(max_milk) %>%
  group_by(Region) %>%
  mutate(Bin = bin(Total_kL)) %>%
  select(Region, Year, Bin)
  
missing_pt <- setdiff(canada_abb, unique(milk_year$Region))

missing_milk <- data.frame(Region = rep(missing_pt, each = 40), 
                           Year = rep(1976:2015, 3),
                           Bin = rep(0, 120), stringsAsFactors = FALSE)
milk_year <- rbind(milk_year, missing_milk)

milk_year$color <- sapply(milk_year$Bin, function(x){
  if(x == 0){
    "grey80"
  } else {
    brewer.pal(11, "PuOr")[x]
  }
})

ani.options(interval = 0.4, ani.width = 600, ani.height = 450)

saveGIF(
  for(i in 1976:2015){
    milkgif <- milk_year[milk_year$Year == i,]
    minicanada(milkgif$Region, pt_colors = milkgif$color, pt_name_cex = 1.5)
    title(main = paste("Canadian Milk Production in", i), line = -1)
  }
)
```

![Canada](https://raw.githubusercontent.com/seankross/minimap/gh-pages/images/canada.gif)


```r
# Legal Status of Gay Marriage Over Time (gif)

library(minimap)
library(animation)

determine_color <- function(status){
  if(status == "bbs")
    "#FFE597"
  else if(status == "nl")
    "#F1F1F0"
  else if (status == "dis")
    "#D0C7B9"
  else if(status == "bbca")
    "#FDC471"
  else
    "#817972"
}

ssm$color <- as.character(sapply(ssm$Status, determine_color))

ani.options(interval = 0.5, ani.width = 600, ani.height = 450)

saveGIF(
  for(i in 1992:2015){
    one_year <- ssm[ssm$Year == i,]
    miniusa(one_year$State, state_colors = one_year$color)
    title(main = paste("Legal Status of Same Sex Marriage in", i), line = -1)
  }
)
```

![USA](https://raw.githubusercontent.com/seankross/minimap/gh-pages/images/usa.gif)

NEWS.md
# minimap 0.1

- First release.


tests/testthat/testusa.R
context("Test miniusa()")

test_that("The USA is drawn correctly", {
  expect_silent(miniusa(usa_abb, 1:51))
  expect_silent(miniusa(usa_abb, 1:51, border_colors = rep("red", 51), 
                        state_names = FALSE))
  expect_silent(miniusa(usa_abb, rep("black", 51), border_colors = rep("white", 51), 
                        state_name_colors = rep("yellow", 51), state_name_cex = .5, 
                        font = "serif"))
})

test_that("miniusa() throws appropriate errors", {
  expect_error(miniusa(c(usa_abb, usa_abb), 1:51))
  expect_error(miniusa(usa_abb[1:10], 1:51))
  expect_error(miniusa(usa_abb, 1:10))
  expect_error(miniusa(usa_abb, 1:51, border_colors = 1:10))
  expect_error(miniusa(usa_abb, 1:51, state_names = NA))
  expect_error(miniusa(usa_abb, 1:51, state_name_colors = 1:10))
})

tests/testthat/testcanada.R
context("Test minicanada()")

test_that("Canada is drawn correctly", {
  expect_silent(minicanada(canada_abb, 1:13))
  expect_silent(minicanada(canada_abb, 1:13, border_colors = rep("red", 13), 
                        pt_names = FALSE))
  expect_silent(minicanada(canada_abb, rep("black", 13), border_colors = rep("white", 13), 
                        pt_name_colors = rep("yellow", 13), pt_name_cex = .5, 
                        font = "serif"))
})

test_that("minicanada() throws appropriate errors", {
  expect_error(miniusa(c(canada_abb[-1], "NU"), 1:13))
  expect_error(minicanada(canada_abb[1:10], 1:13))
  expect_error(minicanada(canada_abb, 1:10))
  expect_error(minicanada(canada_abb, 1:13, border_colors = 1:10))
  expect_error(minicanada(canada_abb, 1:13, pt_names = NA))
  expect_error(minicanada(canada_abb, 1:13, pt_name_colors = 1:10))
})

tests/testthat/testmexico.R
context("Test minimexico()")

test_that("Mexico is drawn correctly", {
  expect_silent(minimexico(mexico_abb, 1:32))
  expect_silent(minimexico(mexico_abb, 1:32, border_colors = rep("red", 32), 
                        estados_names = FALSE))
  expect_silent(minimexico(mexico_abb, rep("black", 32), border_colors = rep("white", 32), 
                        estados_name_colors = rep("yellow", 32), estados_name_cex = .5, 
                        font = "serif"))
})

test_that("minimexico() throws appropriate errors", {
  expect_error(miniusa(c(mexico_abb[-1], "DIF"), 1:32))
  expect_error(minimexico(mexico_abb[1:10], 1:32))
  expect_error(minimexico(mexico_abb, 1:10))
  expect_error(minimexico(mexico_abb, 1:32, border_colors = 1:10))
  expect_error(minimexico(mexico_abb, 1:32, estados_names = NA))
  expect_error(minimexico(mexico_abb, 1:32, estados_name_colors = 1:10))
})

R/miniusa.R
#' Make a tile grid map of the United States of America
#'
#' description
#'
#' @param states description
#' @param state_colors description
#' @param border_colors description
#' @param state_names description
#' @param state_name_colors description
#' @param state_name_cex description
#' @param font description
#'
#' @export
#' @examples
#' \dontrun{
#'  miniusa(c(state.abb, "DC"), 1:51)
#' }
miniusa <- function(states, state_colors, border_colors = rep("white", 51),
                    state_names = TRUE, state_name_colors = rep("white", 51),
                    state_name_cex = 1, font = NULL){

  # Make sure all parameters are specified
  if(any(unlist(lapply(list(
    states, state_colors, border_colors, state_name_colors
  ), length)) != 51)){
    stop("Please make sure parameters contain exactly 51 elements.")
  }

  map <- data.frame(State = c('ME','AK','VT','NH','MA','WA','MT','ND','SD',
                              'MN','WI','MI','NY','CT','RI','OR','ID','WY',
                              'NE','IA','IL','IN','OH','PA','NJ','CA','NV',
                              'UT','CO','KS','MO','KY','WV','DC','MD','DE',
                              'AZ','NM','OK','AR','TN','VA','NC','HI','TX',
                              'LA','MS','AL','GA','SC','FL'),
                    X = c(11,0,9,10,11,1,2,3,4,5,6,7,9,10,11,1,2,3,4,5,6,7,8,
                          9,10,0,1,2,3,4,5,6,7,8,9,10,2,3,4,5,6,7,8,0,3,4,5,
                          6,7,8,7),
                    Y = c(7,6,6,6,6,5,5,5,5,5,5,5,5,5,5,4,4,4,4,4,4,4,4,4,4,3
                          ,3,3,3,3,3,3,3,3,3,3,2,2,2,2,2,2,2,1,1,1,1,1,1,1,0),
                    stringsAsFactors = FALSE)

  # Make sure all states are present
  if(!all(map$State %in% states)){
    stop("It appears some state names are repeated or missing.")
  }

  user_map <- data.frame(State = states, scol = state_colors, bcol =
                           border_colors, sncol = state_name_colors,
                         stringsAsFactors = FALSE)

  map <- merge(map, user_map, by = "State")

  text_ <- NULL
  if(state_names){
    text_ <- map$State
  }

  plotbox(map$X, map$Y, map$scol, map$bcol,
          n_xboxes = 12, n_yboxes = 8, text_ = text_,
          text_col = map$sncol, text_cex = state_name_cex, text_font = font)
}


R/plotbox.R
# @param x X coordinates of boxes to be plotted.
# @param y Y coordinates of boxes to be plotted.
# @param color Color of each box.
# @param n_xboxes Width of plot.
# @param n_yboxes Height of plot.
# @param text_ Text in each box. (Optional)
# @param text_col Color of text in each box. (Optional)

plotbox <- function(x, y, color, border_col, n_xboxes, n_yboxes,
                    text_ = NULL, text_col, text_cex = 1, text_font = NULL){
  # Store old par
  old_xaxs <- par()$xaxs
  old_yaxs <- par()$yaxs

  # Make X and Y axis have precise ends
  par(xaxs = "i", yaxs = "i")

  plot(0,0, col = "white", xlim = c(0, n_xboxes), ylim = c(0, n_yboxes),
       bty = "n", xaxt = "n", yaxt = "n", ann = FALSE, asp = 1)

  for(i in 1:length(x)){
    rect(x[i], y[i], x[i] + 1, y[i] + 1,
         col = color[i],
         border = border_col[i])
    if(!is.null(text_)){
      text(x[i] + .5, y[i] + .5, text_[i], col = text_col[i], cex = text_cex,
           family = text_font)
    }
  }

  # Restore old par
  par(xaxs = old_xaxs, yaxs = old_yaxs)

}


R/minimexico.R
#' Make a tile grid map of Mexico
#'
#' @param estados A vector of Mexican state postal abbreviations. This vector must be
#'  some permutation of \code{mexico_abb}.
#' @param estados_colors A vector of "colors" in the R sense. For example strings
#'  (\code{"blue"}), hex codes (\code{"#D0C7B9"}), etc. The ith color in this
#'  vector will be the color of square that represents the ith element of
#'  \code{estados}.
#' @param border_colors Like \code{estados_colors} but specifying the border
#'  of the square.
#' @param estados_names Should the postal codes for each state be displayed in the
#'  center of the state? The default value is \code{TRUE}.
#' @param estados_name_colors Like \code{estados_colors} but specifying the color
#'  of the text displayed in each state.
#' @param estados_name_cex The size of the text displayed inside of each state.
#' @param font The font of the text displayed inside of each state. The values
#'  \code{"serif"}, \code{"sans"}, and \code{"mono"} are safest to use. Use
#'  other fonts at your own risk. If \code{NULL} a sans-style font will be used.
#'
#' @export
#' @examples
#' \dontrun{
#'  minimexico(mexico_abb, 1:32)
#' }
minimexico <- function(estados, estados_colors, border_colors = rep("white", 32),
                    estados_names = TRUE, estados_name_colors = rep("white", 32),
                    estados_name_cex = 1, font = NULL){

  # Make sure all parameters are specified
  if(any(unlist(lapply(list(
    estados, estados_colors, estados_colors, estados_name_colors
  ), length)) != 32)){
    stop("Please make sure parameters contain exactly 32 elements.")
  }

  map <- data.frame(State = c("BCN", "SON", "CHH", "COA", "NLE",
                              "BCS", "SIN", "DUR", "ZAC", "SLP", "TAM",
                              "NAY", "AGU", "GUA", "QUE", "HID",
                              "JAL", "DIF", "TLA", "YUC",
                              "COL", "MOR", "VER", "TAB", "CAM", "ROO",
                              "MIC", "MEX", "PUE", "GRO", "OAX", "CHP"),
                    X = c(c(0, 2:5), 1:6, 2:6, c(3:5, 10), c(4:6, 9:11), 5:10),
                    Y = c(rep(5, 5), rep(4, 6), rep(3, 5), rep(2, 4), rep(1, 6),
                          rep(0, 6)),
                    stringsAsFactors = FALSE)

  # Make sure all states are present
  if(!all(map$State %in% estados)){
    stop("It appears some state names are repeated or missing.")
  }

  user_map <- data.frame(State = estados, scol = estados_colors, bcol =
                           border_colors, sncol = estados_name_colors,
                         stringsAsFactors = FALSE)

  map <- merge(map, user_map, by = "State")

  text_ <- NULL
  if(estados_names){
    text_ <- map$State
  }

  plotbox(map$X, map$Y, map$scol, map$bcol,
          n_xboxes = 12, n_yboxes = 6, text_ = text_,
          text_col = map$sncol, text_cex = estados_name_cex, text_font = font)
}


R/data.R
#' Same sex marriage in the US
#'
#' Changes in the legality of same sex marriage in the United States over time.
#'
#' @source \url{http://www.nytimes.com/interactive/2015/03/04/us/gay-marriage-state-by-state.html}
#' @format A data frame with columns:
#' \describe{
#'  \item{State}{State Abbreviation}
#'  \item{Status}{Legal status. Either \code{bbs} meaning banned by statute,
#'    \code{nl} meaning not legal, \code{legal}, \code{bbca} meaning banned by
#'    constitutional ammendment, or \code{dis} meaning disputed.}
#'  \item{Year}{Year status went into effect.}
#' }
#' @examples
#' \dontrun{
#'  ssm
#' }
"ssm"

#' Postal Abbreviations for The United States of America
#'
#' @examples
#' \dontrun{
#'  usa_abb
#' }
"usa_abb"

#' Postal Abbreviations for Canada
#'
#' @examples
#' \dontrun{
#'  canada_abb
#' }
"canada_abb"

#' Postal Abbreviations for Mexico
#'
#' @examples
#' \dontrun{
#'  mexico_abb
#' }
"mexico_abb"

#' Production and farm value of maple products in Canada
#'
#' @source Statistics Canada. Table 001-0008 - Production and farm value of
#'  maple products, annual. \url{http://www5.statcan.gc.ca/cansim/}
#' @format A data frame with columns:
#' \describe{
#'  \item{Year}{A value between 1924 and 2015.}
#'  \item{Syrup}{Maple products expressed as syrup, total (in thousands of gallons).}
#'  \item{CAD}{Gross value of maple products (in thousands of Canadian dollars).}
#'  \item{Region}{Postal code abbreviation for territory or province, or
#'    "Canada" representing production for the entire country.}
#' }
#' @examples
#' \dontrun{
#'  maple
#' }
"maple"
